CHANGELOG.md

```diff
@@ -70,6 +70,10 @@ It should be noted that some of these features break backwards compatibility. Se
 
 - Parallelism is improved when running the same target with different arguments in certain cases (e.g. the target uses `WITH DOCKER`).
 - Fixed a log sharing upload-resumption bug
+- Fixed multiple issues with the lexer failing to parse certain characters in shell command substitution (`$()`) and single quoted strings.
+  - Some escaped characters, like `\#`, were failing to parse when used inside shell expressions. Example: `$(echo "a#b#c" | cut -f2 -d\#)`
+  - Some characters, like `#`, were failing to parse when used inside single-quoted strings: Example: `'this is a # string'`
+  - Some invalid shell commands may have been parsed without issue. Example: `$(find . \( -name 'Earthfile'\))` is missing a space before `\)`
```

Comment:
I don't think the example is accurate, because if I create an Earthfile as:

```
VERSION 0.7

test:
  FROM alpine
  RUN mkdir sub
  RUN touch sub/Earthfile
  RUN echo $(find . \( -name 'Earthfile'\))
```

when I run `earthly-v0.7.23 +test` it fails with:

```
               +test | --> RUN echo $(find . \( -name 'Earthfile'\))
               +test | find: unpaired '('
```

however I thought the case you found would cause `find` to work correctly under v0.7.23 but not with this fix.

Reply:
I just tested our root Earthfile with:

```
earthly-v0.7.23 --no-cache +lint-newline-ending
```

and it contains the same `find: unpaired '('` error:

```
+lint-newline-ending | --> RUN export f=ast/tests/empty-targets.earth && if [ "$(tail -c 1 $f)" = "$(printf '\n')" ]; then echo "$f is a special-case test which must not end with a newline."; exit 1; fi
+lint-newline-ending | DOCKERHUB_MIRROR_AUTH_FROM_CLOUD_SECRETS=true
+lint-newline-ending | --> RUN set -e; code=0; for f in $(find . -not -path "./.git/*" -type f \( -iname '*.go' -o -iname 'Earthfile' -o -iname '*.earth' -o -iname '*.md' -o -iname '*.json'\) | grep -v "ast/tests/empty-targets.earth" | grep -v "ast/parser/earth_parser.go" | grep -v "ast/parser/earth_lexer.go" ); do if [ "$(tail -c 2 $f)" == "$(printf '\n\n')" ]; then echo "$f has trailing newlines"; code=1; fi; done; exit $code
+lint-newline-ending | find: unpaired '('
```

could it be that we've just always had this error and never noticed it? If that's the case we could just delete this line.

Reply:
Oops, I added this comment to the wrong thread. Just for completeness:

--

Oh, interesting. It turns out this hasn't been working in main. I noticed this while testing and thought it must have been due to the lexer changes. I'll remove it.

https://github.com/earthly/earthly/actions/runs/7561776859/job/20590798774#step:4:35